### PR TITLE
Minor: Add tests for User Defined Aggregate functions

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

This is part of https://github.com/apache/arrow-datafusion/issues/6611

# Rationale for this change

I am adding support for sliding windows in user defined aggregates and I need some way to test it.  

The actual code change is relatively small but it takes non trivial effort to test, so I wanted to add the tests first so the actual code change is clearer. 

# What changes are included in this PR?

Add more tests to `user_defined_aggregates.rs` for a user defined aggregates. 

# Are these changes tested?
it is only tests

# Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/core/tests/user_defined_aggregates.rs [hunk 1]
```diff
@@ -82,56 +150,151 @@ async fn execute(ctx: &SessionContext, sql: &str) -> Vec<RecordBatch> {
...
+        Self { ctx, counters }
+    }
+}
+
+#[derive(Debug, Default)]
+struct TestCounters {
+    /// was update_batch called?
+    update_batch: AtomicBool,
+    /// was retract batch called?
+    retract_batch: AtomicBool,
```
Comment (stuartcarnie) [inline datafusion/core/tests/user_defined_aggregates.rs:192]: Nice way to test expectations 👌🏻

**Review by stuartcarnie — COMMENTED** [review]
Very nice, @alamb – can I help you by submitting my branch as a PR to this branch?

File: datafusion/core/tests/user_defined_aggregates.rs [hunk 2]
```diff
@@ -82,56 +150,151 @@ async fn execute(ctx: &SessionContext, sql: &str) -> Vec<RecordBatch> {
...
+    fn new() -> Self {
+        Default::default()
+    }
+
+    /// Has `update_batch` been called?
+    fn update_batch(&self) -> bool {
+        self.update_batch.load(Ordering::SeqCst)
+    }
+
+    /// Has `update_batch` been called?
```
Comment (viirya) [inline datafusion/core/tests/user_defined_aggregates.rs:205]:
```suggestion
    /// Has `retract_batch` been called?
```

File: datafusion/core/tests/user_defined_aggregates.rs [hunk 3]
```diff
@@ -82,56 +150,151 @@ async fn execute(ctx: &SessionContext, sql: &str) -> Vec<RecordBatch> {
...
-    ctx
+        Self { ctx, counters }
+    }
+}
+
+#[derive(Debug, Default)]
+struct TestCounters {
+    /// was update_batch called?
+    update_batch: AtomicBool,
+    /// was retract batch called?
```
Comment (viirya) [inline datafusion/core/tests/user_defined_aggregates.rs:191]:
```suggestion
    /// was retract_batch called?
```

**Review by viirya — APPROVED** [review]

**Comment by alamb** [comment]
> Very nice, @alamb – can I help you by submitting my branch as a PR to this branch?

Thanks for the offer @stuartcarnie ! I already incorporated the code from your branch into https://github.com/apache/arrow-datafusion/pull/6671 (which I will have up as a PR shortly)

**Pull request merged**